### DISCUSSION TITLE
Input type checkbox won't call onChange on first interaction when using localStorage

### DISCUSSION THREAD

**Discussion by 0xskinnyg** [vercel/next.js#42892] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:29 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T8101
    Binaries:
      Node: 18.8.0
      npm: 8.18.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

Brave Version 1.45.123 Chromium: 107.0.5304.110 (Official Build) (arm64)

### Describe the Bug

The controlled input type checkbox won't trigger `onChange` when the `checked` prop is set by a state hook initialized from localStorage data.

When I initialize a boolean state with data from localStorage with the following code:

```
useState<boolean>(() => {
  if (typeof window !== 'undefined') {
    const value = localStorage.getItem('auto-connect')
  
    return JSON.parse(value)
  }
```

And provide the following function to the onChange prop:

```
 <input type="checkbox" checked={autoConnect} onChange={() => setAutoConnect((prevState) => !prevState)} />
```

For some reason, when the localStorage value is `true`, the `onChange` won't trigger the function on the first page render and after hard refreshing. From what I see, it initializes the `input` element with `false` before setting the state, but when I check on the DevTools/Components, the object has the `true` value: 
```
[
  {
    "name": "AutoConnect",
    "subHooks": [
      {
        "name": "Context",
        "value": {
          "autoConnect": true,
          "setAutoConnect": "ƒ bound dispatchAction() {}"
        },
        ...,
      }
    ],
    ...,
  }
]
```

Also on the regular DevTools element:
```
<input type="checkbox" class="toggle" checked="">
```

if I don't initialize the state using the localStorage value and just return true, everything works well. 

### Expected Behavior

The controlled input type checkbox should trigger the `onChange` function when interacting with it.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/r1ck12/solana-dapp-next/tree/fix/auto-connect

### To Reproduce

- Clone the linked repository
- Run yarn install
- Run yarn upgrade @solana/web3.js --latest (To fix a bug not pushed yet)
- Run yarn dev
- Access [the localhost](http://localhost:3000/)
- Click on the settings icon
- If the auto-connect is set to true, click on it and the component state won't change

**Comment by icyJoseph** [comment]
I think this is a hydration bug. The HTML sent by the server to the client, doesn't have a checked input element, but the client side rendered element does. Oddly enough `react-dom` is not throwing error upon finding the divergence. 

So there's a compromise to be made, and this is often a pitfall of migrating from CSR to SSR, that many overlook, the server initial frame cannot have any idea of the local storage value, which leaves the client open to diverge and cause hydration issues, in this case, misplacement of a listener.

## Option 1

In this case the least change option is to always start with false, and update it ASAP from the client.

- This might cause corruption in other areas where the auto-connect is consumed once, for example you might report to analytics tools that there was no auto-connect, followed by an event that say there's auto-connect, or you might show tooltip telling the user to auto-connect, only to bring it down immediately. 
- To prevent the above, we could stop using a boolean and use a string union `"pending" | "active" | "disabled"`, and base our logic upon these values.
- useLayoutEffect might look like a great idea but with SSR it becomes a noop on the server, and might cause even more issues

```tsx
export const AutoConnectProvider = ({ children }: { children: ReactNode }) => {
  const [autoConnect, setAutoConnect] = useState(false);

  useEffect(() => {
    const value = localStorage.getItem("auto-connect");
    const initial = value ? JSON.parse(value) : false;

    setAutoConnect(initial);
  }, []);

  const callback = useCallback<Dispatch<SetStateAction<boolean>>>((next) => {
    setAutoConnect((prev) => {
      const update = typeof next === "boolean" ? next : next(prev);
      localStorage.setItem("auto-connect", JSON.stringify(update));

      return update;
    });
  }, []);

  return (
    <AutoConnectContext.Provider
      value={{ autoConnect, setAutoConnect: callback }}
    >
      {children}
    </AutoConnectContext.Provider>
  );
};
```

## Option 2

Use a JS available `document.cookie`, which you can read on the server

- This implies that you'd generate the pages on demand, through something like GSSP or getInitialProps in `_app`
- You'd render the correct UI at all times, but at the cost of not being able to pre-render, or at best, pre-render with caveats (no cookies during build time)


A final consideration is that localStorage can be mutated without your knowledge, so relying on it is often not the best approach. It is almost better to read to use it as a way to preserve data between sessions rather than constantly reading from it during the life time of the application.
